src/Control/Effect/Empty.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE DeriveFunctor, DeriveGeneric, FlexibleContexts, FlexibleInstances, KindSignatures, MultiParamTypeClasses, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Empty
+( -- * Empty effect
+  Empty(..)
+  -- * Empty carrier
+, runEmpty
+, EmptyC(..)
+  -- * Re-exports
+, Carrier
+, Member
+, run
+) where
+
+import Control.Applicative (Alternative (..), liftA2)
+import Control.Effect.Carrier
+import Control.Monad (MonadPlus (..))
+import qualified Control.Monad.Fail as Fail
+import Control.Monad.Fix
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import GHC.Generics (Generic1)
+
+data Empty (m :: * -> *) k = Empty
+  deriving (Functor, Generic1)
+
+instance HFunctor Empty
+instance Effect   Empty
+
+
+-- | Run an 'Empty' effect, returning 'Nothing' for empty computations, or 'Just' the result otherwise.
+--
+--   prop> run (runError empty)    === Nothing
+--   prop> run (runError (pure a)) === Just a
+runEmpty :: EmptyC m a -> m (Maybe a)
+runEmpty = runEmptyC
+
+newtype EmptyC m a = EmptyC { runEmptyC :: m (Maybe a) }
+  deriving (Functor)
+
+instance Applicative m => Applicative (EmptyC m) where
+  pure = EmptyC . pure . Just
+  EmptyC f <*> EmptyC a = EmptyC (liftA2 (<*>) f a)
+
+instance Applicative m => Alternative (EmptyC m) where
+  empty = EmptyC (pure Nothing)
+  EmptyC a <|> EmptyC b = EmptyC (liftA2 (<|>) a b)
+
+instance Monad m => Monad (EmptyC m) where
+  EmptyC a >>= f = EmptyC (a >>= maybe (pure Nothing) (runEmptyC . f))
+
+instance Fail.MonadFail m => Fail.MonadFail (EmptyC m) where
+  fail = lift . Fail.fail
```

Comment:
I opted not to implement `fail` as a synonym for `empty`, since it would lose information (the error message), and since IMO it would be surprising to have `FailC` in a stack under `EmptyC` and have `fail` silently truncate the computation.

Reply:
I wonder if we can use a custom type error message to suggest `FailC` instead.

Reply:
That would rule out absolutely any use of `MonadFail` where `EmptyC` wraps the `Fail` carrier. As written, this allows `MonadFail` to be used with e.g. `EmptyC (FailC PureC)`, which seems a lot friendlier.